### ISSUE TITLE
Hot or Not Game Backend

### ISSUE THREAD

**Issue by ravi-sawlani-yral** [dolr-ai/yral-backend-canister#331]
backend 

- [x] #338
- [x] #340
- [x] #342
- [x] #341

**Comment by abeeshake-yral** [comment]
Here are the findings for the approach for infinite slots in hot or not game.

#### DATA structures: 

```
// system time is the time when first best placed.
map = BTreeMap<PostId, SystemTime>  

queue = Vec<PostId> 


both BTreeMap and Vec are from ic_stable_structures
```


Here is the flow for a bet.
1. when user places a bet 
a) check if the post_id exists in `map`. 
b) if it does, ignore.
c) if it does not, put post_id in `map` and push it at end of `queue`. 

Current algo for betting slot on a post starts the slot after the first bet comes. So, the slot for each post is dynamic. Slot starts right after the first bet comes and slot lasts for 60 minutes.  Slots are not linked to post created_at time to ensure the fairness in the game. 

2. Timers on the bet
a) When user places a bet, we run the timer if not running already. 

b) For a canister, with zero posts in `queue` => When the first bet comes, the timer is started and relevant changes are made to `queue` and `map`. the timer is stored in  a `global_timer` variable. 

c) For canister with some posts in `queue` => new bet comes => Check if the `global_timer` is running (i.e. has Some(TimerId)). If it is running, don't start a new timer. Just add the bet  in `map` and `queue`.

d) when the old timer finishes, it evaluates the `map` and `queue` to setup another timer.

**Comment by abeeshake-yral** [comment]
findings 2024-07-09 

- room ids are tied closely with slots. So, we need slot_ids to determine room_ids 
- Slot_ids is u8. Since, slots are infinite, slot numbers will not fit in u8. 
- Two ways to approach this :
- either use u32 slots or continue with u8 and let overflow happen with the slots. (this will overwrite the slot data every 10-11 days).

**Comment by abeeshake-yral** [comment]
https://github.com/TwistingTwists/test_u8_upgrade

**Comment by abeeshake-yral** [comment]
Here is a  list of sub-tasks for the game. Will check them as we progress. 

- [x] test upgrade of slot_id in separate repo
- [ ] New slots are used correctly in hot or not backend canister
- [ ] hot or not game simulation test
  - [ ] re enqueue timer on upgrades

**Comment by harshitasrivastav28** [comment]
-[] Event instrumentation of HON game

**Comment by harshitasrivastav28** [comment]
- [ ] Event instrumentation

**Comment by harshitasrivastav28** [comment]
- UAt for the app can be available by Wednesday 17th Jul

**Comment by harshitasrivastav28** [comment]
@abeeshake-yral Please comment on the timeline for UAT and delay expected in go-live

**Comment by harshitasrivastav28** [comment]
Expected to be finished today in first half!

**Comment by Natasha-GB** [comment]
- # Entries to be read/inserted to be analysed
- Decide approach based on the above
- Planned for closure by today
_Vector to store number of posts user will be betting on_

**Comment by harshitasrivastav28** [comment]
-Ensure backward compatibility of the end points

**Comment by abeeshake-yral** [comment]
UPDATED : 
https://github.com/go-bazzinga/hot-or-not-backend-canister/issues/341#issuecomment-2251905756

TODO: 

- frontend backwards compatibility 

1.  `did` files changed. Approved by Ravi.  [vec nat8 -> blob](https://github.com/go-bazzinga/hot-or-not-backend-canister/pull/330/files#diff-eca1ba756b75500397d13998ffc553b114c4d17035f8fade49827425182571d1R13)
2. nat64 for slot_id in frontend structs :  https://github.com/go-bazzinga/hot-or-not-backend-canister/pull/330/files#r1691183127

- backend backwards compatibilty -- 
1.  ([integration test to check the upgradation from SlotId to NewSlotType to avoid canister crashes](https://github.com/go-bazzinga/hot-or-not-backend-canister/pull/330#pullrequestreview-2197157156)) 
2. [host or not game simulation_2 in local](https://github.com/go-bazzinga/hot-or-not-backend-canister/pull/330#discussion_r1690114704)
3. possible integration test -->  Old posts with `hot_or_not_details` = `None` - [can we play HON game on them? ](https://github.com/go-bazzinga/hot-or-not-backend-canister/pull/330#pullrequestreview-2198288624)

**Comment by Natasha-GB** [comment]
- Integration test for BE (1)
- BE (2&3) will be done today - might spill to first half tomorrow

**Comment by abeeshake-yral** [comment]
First PR is up : https://github.com/go-bazzinga/hot-or-not-backend-canister/pull/352 

Second PR is in progress. Will be raised today. 

BBC - 3. in progress. expected to be done by tomorrow.

**Comment by abeeshake-yral** [comment]
Second PR is up: #354  

With second PR , Here is what is finished 
1. frontend and backwards compatibility 
2. integration tests are in place. 

While this is being reviewed, I will move on to finishing frontend tasks.

**Comment by Natasha-GB** [comment]
- Remove the feed check and use old endpoints to ship faster
- Create separate task to implement the timer part (infinite instead of 48 hours)

**Comment by abeeshake-yral** [comment]
Betting starts when the first user bets on a given post. Betting slot is alloted thena and a timer for 1 hr is started.


## Here are two apporaches to implement the timer:

1. have a timer for each post which is being betted on
2. have a global timer and a queue of posts that are being betted on.
   1. iterate over the queue and check if the timer has expired.
   2. if the timer has expired, then tabulate the outcome for the post and remove the post from the queue.

We chose the second approach because it is more efficient.

### Here are the scenarios that can happen for the timers:

post_2_Z => means post 2 for the user with id Z

1. User A bets on a post_1_Z at 04:00  and the global timer for 1 hr is started (to be finished at 05:00)
2. User B bets on a post_1_Z at 04:01 and no timer is started. (since there should be ONLY one timer in the canister at a time)
3. User C bets on a post_2_Z at 04:02 and no timer is started. (since there should be ONLY one timer in the canister at a time)


When timer expires, post_1_Z is `tabulate_hot_or_not_outcome_for_post_slot_v1` and post_1_Z is removed from the queue. Next timer is queued for the next post (post_2_Z) in queue. 

Note that this second timer actually needs to run only for 1 minute since this timer is started at 05:00 and the expiry of this timer needs to be at 05:01.  Computationally, the function `tabulate_hot_or_not_outcome_for_post_slot_v1` after 1 hr for post_2_Z. 


### Here are the technical details for implementing the timer:

1. queue is best implemented in stable memory as `ic_stable_structures::btree_map::BTreeMap` because it has `pop_first` and `insert` methods. The Vec in ic_stable_structures doesn't have `pop_first`.
2. we also need to keep track of the time when the first bet was placed for each post. This is because timer expires 1 hr after the first bet on the given post is placed.
```rust
    pub bet_timer_posts: ic_stable_structures::btreemap::BTreeMap<(SystemTimeInMs, PostId), (), Memory>,
```

#### Here are the scenarios that can happen for the timers:

Think of these sequentially.

A1. when first bet on post_1_Z is placed, and bet_timer_posts is empty. We insert into bet_timer_posts a tuple of (SystemTimeInMs, PostId) and `start_timer`.

A2. when second bet on post_1_Z is placed, we *check* if PostId(post_1_Z) exists in bet_timer_posts. if it doesn't, then we insert into bet_timer_posts a tuple of (SystemTimeInMs, PostId). we don't ned to call `start_timer` here. Since there must be a global timer running. It was set in A1. 

A3. when first bet on post_2_Z is placed, we *check* if PostId(post_2_Z) exists in bet_timer_posts. It doesn't. So, we insert into bet_timer_posts a tuple of (SystemTimeInMs, PostId). we don't ned to call `start_timer` here.

A4. when timer for post_1_Z expires, we `tabulate_hot_or_not_outcome_for_post_slot_v1` for post_1_Z and remove it from bet_timer_posts. also, start the next timer (via `start_timer`) for the next post (post_2_Z) in the queue bet_timer_posts.

Now, in A2 above, to check if the posts has already been betted upon, in O(1) time, we need a separate BTreeMap<PostId, ()>. we call this map `first_bet_placed_at_hashmap`.
now this map needs to be updated always along with bet_timer_posts. at the cost of memory, we are optimising for lookup speed.


originally posted : https://github.com/go-bazzinga/hot-or-not-backend-canister/pull/354#issuecomment-2265673560

**Comment by Natasha-GB** [comment]
- Issue fixed related to instructions exceeded (@komal-rs's update) Slotid made u64 and Roomid made u16 - all u64s are causing an issue <not a blocker>

**Comment by abeeshake-yral** [comment]
Komal reviewed the PR. Suggested a way to merge PR in one step. Testing that right now.

**Comment by Natasha-GB** [comment]
- 2 steps left to try out the process (should be done in less than an hour)
- Integration can be done today

**Comment by harshitasrivastav28** [comment]
- Backend PR has been merged the v1
- Merging the timer part which has been fixed by Komal is pending

**Comment by harshitasrivastav28** [comment]
- Ensure the current frontend
- Ensure the migration we are able to do 
- Facing panic in main - Using docker for now
- WASM exceeded call issue is not resolved

**Comment by abeeshake-yral** [comment]
Blocked on call exceeded issue. 

- it is not amd vs arm issue because Adarsh has linux x86_64 and he gets a different  error than "wasm call exceeded". 
- cannot test for the code on arm linux because ic-nix  binaries are not available for arm linux. They are available for arm MacOs

Next step for this is: 
- figure out a way to make it run with Adarsh's laptop. We both are using nix-shell. 

Error on Adarsh's laptop: 
![image](https://github.com/user-attachments/assets/38e941d3-7fca-4d25-b4c0-65b6bf87a8d1)

**Comment by harshitasrivastav28** [comment]
- Discuss with @komal-rs and @rupansh and get the infinite timer closed.

**Comment by harshitasrivastav28** [comment]
- @abeeshake-yral to deploy from his account and check if this is working

**Comment by harshitasrivastav28** [comment]
- We were deploying that on mainet. Our backend deployment is blocked on this itself.
- Suggestion is to re-deploy it. If we can get subnet status to understand if subnet was off that day
- Do selective upgrade on user indexes 
- @ravibazz to oversight this project. 
- Check if there are specific subnet which have not been upgraded?
